medusa/restore_cluster.py

```diff
@@ -222,23 +224,13 @@ def _chunk(my_list, nb_chunks):
                 topology_matches = False
 
         if topology_matches:
-            # We can associate each restore node with exactly one backup node
-            backup_ringmap = collections.defaultdict(list)
-            target_ringmap = collections.defaultdict(list)
-            for token, host in backup_tokens.items():
-                backup_ringmap[token].append(host)
-            for token, host in target_tokens.items():
-                target_ringmap[token].append(host)
-
-            self.ringmap = backup_ringmap
-            i = 0
-            for token, hosts in backup_ringmap.items():
-                # take the node that has the same token list or pick the one with the same position in the map.
-                restore_host = target_ringmap.get(token, list(target_ringmap.values())[i])[0]
+            # backup and restore nodes are ordered by smallest token and associated one by one
```

Comment:
I don't understand why it's necessary to sort the nodes considering target nodes are going to get completely new token assignments.

Reply:
We need a common way of associating nodes, that would give the exact same result when invoked from each pod in the target cluster.
If it's an in place restore, nodes will be associated with themselves (or the node that they replaced). If it's a remote restore, they'll be associated with one of the nodes in a consistent way.
The tokenmaps are not ordered, which made the mapping inconsistent before I did the above change. I would get some pods to map themselves to the same backup node. This is different to what Medusa itself does for restore orchestration as the mapping is done just once when `medusa restore-cluster` is invoked. In the context of k8s, it's invoked once per pod, which requires consistency in the results.

Reply:
My understanding based on your prior explanations is that the token mappings between  src and dest clusters is done to determine whether it is an in-place or remote restore. In the context of k8ssandra this means that the restore cluster has to be created with all Cassandra pods reaching the ready state and then a restart of all pods to perform the actual restore. 

When used outside of k8s this makes sense since Medusa doesn't really have a true orchestration layer. In k8ssandra-operator though, we do have that orchestration layer. k8ssandra-operator can easily tell Medusa if it is a remote restore. If the operator tells Medusa, is the mapping necessary? If not it would simplify things quite a bit as well as make for a better experience for the user.

Reply:
> k8ssandra-operator can easily tell Medusa if it is a remote restore

How so?
k8ssandra-operator would need to check if the topologies match, requiring it to do more work than it does now.

> If the operator tells Medusa, is the mapping necessary? If not it would simplify things quite a bit as well as make for a better experience for the user.

I don't understand what you see that would be simpler than this. The mapping is done automatically by Medusa and the user doesn't need to distinguish in place and remote restores, they're both handled seamlessly as long as you have the same number of nodes.
Medusa needs the mapping to be computed internally anyway for non k8s setups, so this code is necessary. Re-using it seems like a better option than duplicating the logic in the operator.

Reply:
and we'll soon add more safety around rack assignments, which will be much more easily done by Medusa itself.

Reply:
If I understood this correctly, this makes a horrible user experience if everything isn't automated? First the user has to create a K8ssandraCluster and then start the restore process? So the operator might have to be there waiting say an hour for the first process to finish and only then move forward with the operation? To me that feels like an ugly hack and not very Kubernetes-native. That's not what we should be shipping. Lets ignore for a moment the fact how the restore process is done underneath, to user that isn't visible.

What we really should have is a single CRD I create that takes care of everything, including the whole underneath ugly restore processes, starting up correct size Cassandra cluster and giving me the data I originally had. Whether we do this by storing K8ssandraCluster CRD somewhere or calculate the required nodes or make a guess is irrelevant in the first phase. The important thing is that it's a simple process where the operator can simply do one command, go to sleep and come back next morning with an operational cluster. 

Because not only does it reduce the amount of load on the human operator (equals less cost), it also for example allows simple automated testing with the remote restore. I could set my CI/CD job create a staging environment using the latest production data, run the tests and then tear down the setup. And it should all be simple to do. If it requires me to build automation to do the medusa part, then we should do the automation part in to our software. 

There are other benefits in using the operator calculated statistics also. We could be looking at the running Kubernetes cluster and check if there are even resources to load the backup there. Is there enough resources available, enough nodes available, we could do modified Kubernetes scheduling with operator, or we could recalculate new optimal size for cluster (if we could restore to different size - out of this PR, but for future) depending on the Kubernetes cluster size. We could even push for a HPA rule to spin up new nodes in managed Kubernetes cluster to actually get enough resources. Kubernetes is a building block we should leverage and allow us to grow, not limit ourselves to limitations of the old.

If the non-Kubernetes version becomes a burden, we should really consider forking the project to two and not limit our possibilities, just because someone wants to avoid few lines of code.

Reply:
> this makes a horrible user experience

Well that's not excessive at all 😂 

> What we really should have is a single CRD I create that takes care of everything, including the whole underneath ugly restore processes, starting up correct size Cassandra cluster and giving me the data I originally had. Whether we do this by storing K8ssandraCluster CRD somewhere or calculate the required nodes or make a guess is irrelevant in the first phase. The important thing is that it's a simple process where the operator can simply do one command, go to sleep and come back next morning with an operational cluster.

I'm sure we can build some additional automation on top of what we have in this PR, but I'd like to make that a separate follow up work. There are lots of changes already here and they match the scope that's been designed.

> If the non-Kubernetes version becomes a burden, we should really consider forking the project to two and not limit our possibilities, just because someone wants to avoid few lines of code.

We're not doing it this way out of laziness, there are more reasons to that 🙂 And what you're describing makes that Kubernetes implementation incompatible with backups taken with Medusa on non k8s Cassandra clusters, which I think is a good feature to have.

Reply:
> There are lots of changes already here and they match the scope that's been designed.

I disagree with the definition "scope that's been designed". I don't think we ever agreed on something that's not functional in Kubernetes. Having separate steps which is all due to the fact that this isn't Kubernetes friendly implementation was never the design target. The design specs talked about Helm, which has restrictions on how to create those resources - but those restrictions do not exists with k8ssandra-operator. 

We should rather have a version that doesn't work standalone for restoring than one that doesn't work correctly in Kube. I don't agree with follow-up PR, unless the other PR is ready before this PR is merged. This PR is only few hundred lines, so it's not that big (and some of those lines is generated code).

> We're not doing it this way out of laziness, there are more reasons to that

The previous comment of yours mentioned "duplicate lines" and such, that's not a very good reason in this case when it makes the functionality a lot better. And which it would. 

There's no good UI reason why we wouldn't do this the correct way if we introduce the feature. If there are limitations that prevent this from happening in Kube and non-Kube ends, then we should go with the Kube-only implementation.

> And what you're describing makes that Kubernetes implementation incompatible with backups taken with Medusa on non k8s Cassandra clusters, which I think is a good feature to have.

I didn't say that, the different restore process shouldn't be something that breaks the backup taking capability. A separate future with most work done in Kube-only-Medusa? That could still support older versions (or update path), that shouldn't be an issue either even if it eventually would separate to non-compatible form (eg. use 3.x to remote-restore 3.x backups and 4.x supports only 4.x backups etc). 

The matter of the fact is, we can't have limitations in features just because we need to support legacy. Compromising to both directions is not a good way forward, that has to be cut at some point. "Fork" is just a project management way where we would get rid of the stuff that's not necessary inside Kubernetes and allow us to do stuff that wouldn't work outside without Kubernetes. It would probably even simplify the codebase quite a bit.